ibc/docs/ics23.md

```diff
@@ -0,0 +1,78 @@
+
+## Implementation
+
+The benefit of using `cosmos/ics23` over implementing similar types ourselves is twofold:
+
+1. It is already SDK agnostic, so can be used by Pocket (a non-cosmos chain) without any issues.
+2. The functions defined for proof verification are separate from the tree structures themselves
+   - This means we do not need to interact with an SMT instance in order to verify proofs.
+
+However, there were some changes made specifically for Pocket's implementation of ICS-23.
+
+### Custom SMT `ProofSpec`
+
+The `ProofSpec` type in `cosmos/ics23` is used to define the steps needed to verify a proof, what hashing functions should be used, any node prefixes, etc. This is then passed into the verification functions in order to verify a proof, instead of having to interact with the tree itself. This is useful as proofs must be verified on a light-client and as such being able to verify a proof without creating a tree is much more memory efficient.
+
+As the SMT used by Pocket Network only stores hashed values by default, the IBC stores use the `WithValueHasher(nil)` option which stores the raw bytes of the values in the tree. As such the following `ProofSpec` was created:
+
+```go
+smtSpec *ics23.ProofSpec = &ics23.ProofSpec{
+	LeafSpec: &ics23.LeafOp{
+		Hash:         ics23.HashOp_SHA256,
+		PrehashKey:   ics23.HashOp_SHA256,
+		PrehashValue: ics23.HashOp_NO_HASH,
+		Length:       ics23.LengthOp_NO_PREFIX,
+        Prefix:       []byte{0},
+	},
+	InnerSpec: &ics23.InnerSpec{
+		ChildOrder:      []int32{0, 1},
+		ChildSize:       32,
+		MinPrefixLength: 1,
+		MaxPrefixLength: 1,
+        EmptyChild:      make([]byte, 32),
+		Hash:            ics23.HashOp_SHA256,
+	},
+	MaxDepth:                   256,
+	PrehashKeyBeforeComparison: true,
+}
+```
+
+The main change from the `cosmos/ics23` `SmtSpec` object is that the `PrehashValue` field is set to not hash values prior to hashing the key-value pair.
+
+### Converting `SparseMerkleProof` to `CommitmentProof`
+
+In order to convert the proofs generated by the SMT into a serialisable proof used by `cosmos/ics23`, the `SideNodes` field of the `SparseMerkleProof` must be converted into a list of `InnerOp` types which define the order of the hashes. The order of the hashes is important as depending on whether the next hash is the left or right neighbour of the current hash, they will be hashed in a different order, ultimately creating a different root hash. This conversion allows the verification to produce the same root hash as the SMT would have produced when verifying the proof.
```

Comment:
There were old discussions in celestia's repo on this. It is possible and we could do it. It would be a heft task though. A cool idea would be setting `k` on creation. But as stated in those previous discussions we would need to benchmark it appropriately.

I am thinking of creating a suite of benchmark tests in the SMT we could then use as a baseline for any changes moving forward

Reply:
Yeah I'd like to explore this, as I mentioned it was a topic of discussion in the original SMT repo and would be cool to find our own optimisation - potentially for different node stores we could provide a optimal `k` constant:
```go
const (
  SimpleMapStoreK = 2
  BadgerK = 4
)
```
This could be cool.